lib/os/mempool.c

```diff
@@ -236,67 +231,72 @@ static void *block_break(struct sys_mem_pool_base *p, void *block, int l,
 int z_sys_mem_pool_block_alloc(struct sys_mem_pool_base *p, size_t size,
 			      u32_t *level_p, u32_t *block_p, void **data_p)
 {
-	int i, from_l, alloc_l = -1, free_l = -1;
-	unsigned int key;
-	void *data = NULL;
+	/* Array of sizes for each level. Populated on demand to save RAM. */
 	size_t lsizes[LVL_ARRAY_SZ(p->n_levels)];
-
-	/* Walk down through levels, finding the one from which we
-	 * want to allocate and the smallest one with a free entry
-	 * from which we can split an allocation if needed.  Along the
-	 * way, we populate an array of sizes for each level so we
-	 * don't need to waste RAM storing it.
-	 */
 	lsizes[0] = _ALIGN4(p->max_sz);
-	for (i = 0; i < p->n_levels; i++) {
+
+	/* Find level from which we want to allocate. */
+	int alloc_l = -1;
+	for (int i = 0; i < p->n_levels; i++) {
 		if (i > 0) {
-			lsizes[i] = _ALIGN4(lsizes[i-1] / 4);
+			lsizes[i] = _ALIGN4(lsizes[i - 1] / 4);
 		}
 
 		if (lsizes[i] < size) {
 			break;
 		}
 
 		alloc_l = i;
-		if (!level_empty(p, i)) {
-			free_l = i;
+	}
+
+	if (alloc_l < 0) {
+		*data_p = NULL;
+		return -ENOMEM;
+	}
+
+	void *data = NULL;
+
+	unsigned int key = pool_irq_lock(p);
+
+	/* Try allocating the smallest block possible.
+	 * Search must be performed atomically to make sure block rearranging
```

Comment:
The existing design was intended to prevent exactly this circumstance, though.  Making the allocation fully atomic is a latency bug.

Reply:
The only thing added is checking if level is empty, these are just few operations. Note that `block_alloc` will return immediately if level is empty. Only when there is block available on a level checked `block_alloc` will do something expensive.
So you will end up with once expensive `block_alloc` and few `block_break` (if needed) - both already under a lock.
I personally prefer that to unbound operation with repetitive retries.